Rewrite lintElses check to handle multiple else ifs

zimmski opened the pull request:
As suggested by @dsymonds this PR handles multiple else-ifs in lintElses. The suggestion for outdenting an else block is only shown if the if block has a return as last statement, no else-ifs and of course an else block.The PR also adds some test cases to make the behavior clearer.

zimmski committed:
Fix, recommend outdenting of else blocks only if there are no else-ifs

dsymonds commented on lint.go:
```diff
@@ -991,46 +991,63 @@ func (f *file) lintVarDecls() {
 	// They will appear as an IfStmt whose Else field is also an IfStmt.
 	// Record such a node so we ignore it when we visit it.
-	ignore := make(map[*ast.IfStmt]bool)
+	ignore := make(map[*ast.IfStmt]struct{})
```
revert. `map[T]bool` is the most common way to represent a set, and the space savings for using `map[T]struct{}` are smaller than people tend to assume.

zimmski replied:
done

dsymonds commented:
The code here is meant to completely ignore `else if` blocks, mostly because their control flow is a lot harder to see than just if+else. Can you make this change correctly ignore an `else if` chain, rather than trying to flag the dangling `else` when all the previous blocks return?

zimmski commented:
@dsnet: I guess this could be done much cleaner but it addresses the concerns of dsymonds without changing too much of the original PR. Please take a look.

zimmski commented:
@dsnet: I resolved @dsymonds concerns and adapted the whole PR to be smaller since most cases are not needed anymore because of the behavior change. Please take a look.

zimmski commented:
@dsnet Any update on this?

JKobyP commented:
Any updates on this?

zimmski commented:
@dsymonds since this came up with #354 again. Could you please take another look at this PR? I would be really sad if the invested time would be wasted. (I guess since it is that old a CI rerun would be also a good idea)

andybons commented:
@alandonovan can you take a look at this?

alandonovan commented on lint.go:
```diff
@@ -1027,34 +1027,47 @@ func (f *file) lintElses() {
+
+		outdentElse := true
+
+		// The else-block should only be outdented if the if block ends with a return statement.
```
Is there are reason to detect 'return' but not other control statements such as 'continue' and 'break'?

zimmski replied:
The original code handles only "returns". I guess it would be better to add "continue" and "break" in another PR unless you really want me to?

alandonovan commented on lint.go:
```diff
@@ -1027,34 +1027,47 @@ func (f *file) lintElses() {
-		if _, ok := ifStmt.Else.(*ast.BlockStmt); !ok {
-			// only care about elses without conditions
+
+		if _, ok := ignore[ifStmt]; ok {
```
It's quite normal to use a map[T]bool containing only true values to represent a set of T, and idiomatic to test for membership of such a set using just m[k].  Please restore the previous check.

alandonovan reviewed (approved):
Code looks fine to me, but I'm not really an owner of this project.

andybons commented:
@zimmski ping

andybons commented:
@googlebot cla?

zimmski commented:
Funny enough, I did not received any notifications for @alandonovan comments but got @andybons comments right away? So just in case my comments for the code are not going through:
a.) I made the one change @alandonovan requested, rebased to the latest master and pushed the newest version.
b.) Regarding adding support for "continue" and "break" statements: The original intention of this PR was to fix the else-if problems. Adding these statements is fine with me, but if it is up to me, I would add them in another PR.
